### ISSUE TITLE
On Rebalance, Consumers won't be move till next message?

### ISSUE THREAD

**Issue by steve-gray** [oleksiyk/kafka#103]
One of the interesting issues I've seen is that when a consumer group member starts watching a partition - it may not actually start working against that partition until the next message is sent - even if the partition has data.

i.e.  Partition X - End offset = 256

Consumer joins with startingOffset -2 - at position 0. It won't then read the 256 pre-existing messages until we've pushed a message #257 into the queue.

Is this a setting that's controlled via a flag?

**Comment by oleksiyk** [comment]
Can you replicate this? Is there full output log?

**Comment by steve-gray** [comment]
Do you just need the logs from this with NSL_LEVEL=5?  Or something else?

I can arrange it Monday. Interestingly we've also observed during this scenario that some of the partition consumers don't actually *ever* start. They just keep showing in the Kafka manager as having not started at all.

**Comment by oleksiyk** [comment]
> They just keep showing in the Kafka manager as having not started at all.

Maybe because there are more consumers than partitions?

**Comment by steve-gray** [comment]
100 partitions, issue manifested with 1 consumer and whilst some partitions seemed to come to life as I added a few more consumers in rancher - there was always a few that stayed at -1/not started. (-1 in kafka-manager just means there was no consumed offset stored for that consumer group/partition)

Interestingly if I just blow away the whole group and start the processes anew with a different groupId then it is sometimes ok.

**Comment by steve-gray** [comment]
Here's all I've got from the output with NSL_LEVEL=5

    8/8/2016 8:05:05 AM
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.868Z[0m [38;5;70mINFO[0m worker-engine Joined group worker-group-operations-06 generationId 2 as worker-engine-cb247076-f4d9-49a6-8597-17daef4ef698
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.935Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:67 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.936Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:46 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.936Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:22 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.938Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:94 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.938Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:55 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.938Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:40 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.939Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:79 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.939Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:16 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.939Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:97 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.942Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:76 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.942Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:88 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.943Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:58 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.943Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:82 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.944Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:13 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.944Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:64 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.944Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:91 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.944Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:85 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.945Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:70 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.945Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:52 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.945Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:25 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.946Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:1 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.946Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:49 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.946Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:19 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.946Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:4 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.948Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:31 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.948Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:7 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.948Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:37 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.948Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:34 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.949Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:28 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.950Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:10 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.950Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:43 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.951Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:73 offset 0 leader kafka-events:9092
    8/8/2016 8:05:11 AM[90m2016-08-07T22:05:11.951Z[0m [38;5;24mDEBUG[0m worker-engine Subscribed to events:61 offset 0 leader kafka-events:9092

If I push in a new message to Kafka, *some* partitions start advancing. However, no errors or output to help figure this out.

**Comment by oleksiyk** [comment]
Can it be something with producer partitioner? Maybe it just omits some partitions?

**Comment by steve-gray** [comment]
I thought that too, but the partitioner is a modulo % toHashValue type operation - the distribution is reasonably uniform and all 100 partitions get data in them - something coorobated by the Kafka Manager UI. 

Is there something I can do perhaps to get extra logging behind the consumers and why certain partitions don't start consuming? I might have a play with the code behind the assignment strategies and see if there's something to it - but I believe from eyeballing things that I do at least see all partitions covered in the list of allocations that gets printed after a group rebalance.

**Comment by oleksiyk** [comment]
Try logging out fetch request `data` here: https://github.com/oleksiyk/kafka/blob/master/lib/base_consumer.js#L56 and see if all partitions are there.

**Comment by steve-gray** [comment]
I'll have to figure out a way to hack that into the version of the worker on our hosts where the bulk data is. I'm wondering though: is the concurrency limiter potentially at play here? I'm looking at that code and wondering if it wont process any partitions beyond the first N and because I've got a lot of a partitions - and a lot of messages in them - so the map on line 56 might not run all the partitions at once because it's still grinding through the chain for the other partitions.

Does that make any sense? I'm not in front of a system to check - but it __feels__ plausible. I could have sworn that if I left an instance long enough they all did eventually get processed - but that might have been one someone else restarted whilst I wasn't looking.

**Comment by oleksiyk** [comment]
I don't see how it can be possible unless some of your handlers never return (don't settle).. Next fetch request won't be sent until all data from the previous fetch response is processed with `map`, no matter what concurrency limit is. Try dumping the response too.

**Comment by steve-gray** [comment]
I'm going to be trying this out in more detail tomorrow - I'm thinking to go through sprinking NSL logging code throughout the client/group consumer stuff. Is it worth me putting this stuff through as PR?

**Comment by oleksiyk** [comment]
Probably no need in PR here.

**Comment by steve-gray** [comment]
@oleksiyk: Found the problem. Reproduction scenario:

- Create a Kafka topic with ~100 partitions, use a single broker.
- Put ~100mb of data into each topic in lots of small messages
- Start a new GroupConsumer from the start of the topic (-2)

What happens is that effectively the initial fetch is so punishingly large that it times out things - effectively 100x maxBatchSize - I end up invariably with NodeJS quietly bombing out with a GC error/restarting the process. I'm guessing the times where I observed 'add a message' working was in relation to the passage of time. Production there's 1 partition/broker, so this isn't problematic - but in test with the same partition topology on a single server...

Long story short, changed the maxBytes for the consumer to 64kb - everything seemed much happier. If we'd had the brokers split out as we would in production, I imagine this wouldn't have been much of a problem - since we'd have never got so much data back in a single hit from one broker. Is there perhaps a need here to throttle the number of outstanding requests to a single broker?

**Comment by steve-gray** [comment]
Closing. PEBKAC.
